seahorn/stubs/trusty_messaging_api_override.c

```diff
@@ -0,0 +1,98 @@
+/** Verify Trusty
+
+    Implementation of TrustyOS Messaging API
+    https://source.android.com/security/trusty/trusty-ref#messaging_api
+
+ */
+
+#include "handle_table.h"
+#include "nondet.h"
+#include "sea_mem_helper.h"
+#include "seahorn/seahorn.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+// trusty reference for definitions only
+#include <trusty_ipc.h> // -> ipc structs
+#include <uapi/err.h>   // NO_ERROR definition
+
+static uint32_t cur_msg_id;
+static bool cur_msg_retired = true;
+
+static bool msg_retired(uint32_t msg_id) {
+  return msg_id == cur_msg_id && cur_msg_retired;
+}
+
+// get len of first iov of msg if there is any, otherwise 0
+static size_t msg_capacity(ipc_msg_t *msg) {
+  struct iovec *iovecs = msg->iov;
+  if (msg->num_iov == 0)
+    return 0;
+  return iovecs[0].iov_len;
+}
+
+/** messaging **/
+/** Gets meta-information about the next message in an incoming message queue */
+int get_msg(handle_t handle, ipc_msg_info_t *msg_info) {
+  (void)handle;
+  int retval = nd_get_msg_ret();
+  size_t msg_len = nd_msg_len();
+  uint32_t msg_id = nd_msg_id();
+  if (retval == NO_ERROR) {
+    assume(msg_len > 0);
+    assume(msg_id > 0);
+    cur_msg_id = msg_id;
+    cur_msg_retired = false;
+  } else {
+    msg_len = 0;
+    msg_id = 0;
```

Comment:
why is retired flag left unchanged. what if it was false before

Reply:
this function retrieves meta info of the next message in the internal queue; it should not "retire" the previous message since the only proper way to do that is through `put_msg`; our current implementation kind of under-approximates things, trading preciseness for efficiency